sample/samplers.go

```diff
@@ -1,44 +1,69 @@
 package sample
 
 import (
-	"errors"
+	"math"
 	"math/rand/v2"
 	"slices"
-)
 
-// Sampler is not thread-safe. Each goroutine should have its own instance
-type Sampler interface {
-	Sample([]float32) (int32, error)
-}
+	"github.com/ollama/ollama/llama"
+)
 
-// logit represents information about a single token during sampling
-type logit struct {
+// token represents information about a single token during sampling
+type token struct {
 	id    int32   // The token's unique identifier
 	value float32 // The raw logit or probability from the model
 }
 
-type weighted struct {
+type Sampler struct {
 	rng         *rand.Rand
-	tokens      []logit
 	topK        int
 	topP        float32
 	minP        float32
 	temperature float32
+	grammar     *Grammar
 }
 
-func (s *weighted) Sample(logits []float32) (int32, error) {
-	if len(s.tokens) < len(logits) {
-		s.tokens = make([]logit, len(logits))
+func (s *Sampler) Sample(logits []float32) (int32, error) {
```

Comment:
The sample, check grammar, sample again logic is needed for both greedy and weighted sampling, and so I merged them into one struct for now to avoid having this logic twice.